### PR TITLE
feat (vrf v2 mock): user provided words

### PR DESCRIPTION
**Pull request by cuckflong**
implementation to #6652

### CODE REVIEW

**Comment by cuckflong** [comment]
> Good change! However, some users use this mock file in their existing tests, so this change would break those tests.
> 
> It's best to use overloading:
> 
> ```
> fulfillRandomWords(uint256 _requestId, address _consumer) external {
>   fulfillRandomWords(_requestId, _consumer, new uint256[](0));
> }
> 
> fulfillRandomWords(uint256 _requestId, address _consumer, uint256[] memory _words) external {
>   ...
> }
> ```

Great suggestion! Have updated the code to use overloading

**Comment by cuckflong** [comment]
has updated accordingly

**Review by vreff — APPROVED** [review]

**Comment by vreff** [comment]
Try running this locally:
```
yarn workspace @chainlink/contracts hardhat test
```

**Comment by cuckflong** [comment]
CleanShot 2022-05-27 at 00 47 43@2x

getting the above errors

File: contracts/src/v0.8/mocks/VRFCoordinatorV2Mock.sol [hunk 1]
```diff
@@ -85,19 +85,41 @@ contract VRFCoordinatorV2Mock is VRFCoordinatorV2Interface {
...
+   *
+   * @dev This mock uses a simplified formula for calculating payment amount and gas usage, and does
+   * @dev not account for all edge cases handled in the real VRF coordinator. When making requests
+   * @dev against the real coordinator a small amount of additional LINK is required.
+   *
+   * @param _requestId the request to fulfill
+   * @param _consumer the VRF randomness consumer to send the result to
+   * @param _words user provided random words
+   */
+  function fulfillRandomWords(
```
Comment (vreff) [inline contracts/src/v0.8/mocks/VRFCoordinatorV2Mock.sol:103]:
```suggestion
  function fulfillRandomWordsWithOverride(
```

File: contracts/src/v0.8/mocks/VRFCoordinatorV2Mock.sol [hunk 2]
```diff
@@ -85,19 +85,41 @@ contract VRFCoordinatorV2Mock is VRFCoordinatorV2Interface {
    * @param _consumer the VRF randomness consumer to send the result to
    */
   function fulfillRandomWords(uint256 _requestId, address _consumer) external {
+    fulfillRandomWords(_requestId, _consumer, new uint256[](0));
```
Comment (vreff) [inline contracts/src/v0.8/mocks/VRFCoordinatorV2Mock.sol:88]:
```suggestion
    fulfillRandomWordsWithOverride(_requestId, _consumer, new uint256[](0));
```

File: contracts/test/v0.8/dev/VRFCoordinatorV2Mock.test.ts [hunk 3]
```diff
@@ -230,7 +230,7 @@ describe('VRFCoordinatorV2Mock', () => {
       await expect(
         vrfCoordinatorV2Mock
           .connect(random)
-          .fulfillRandomWords(1, vrfConsumerV2.address),
+          ['fulfillRandomWords(uint256,address)'](1, vrfConsumerV2.address),
```
Comment (vreff) [inline contracts/test/v0.8/dev/VRFCoordinatorV2Mock.test.ts:233]:
```suggestion
          .fulfillRandomWords(1, vrfConsumerV2.address),
```

File: contracts/test/v0.8/dev/VRFCoordinatorV2Mock.test.ts [hunk 4]
```diff
@@ -254,7 +254,7 @@ describe('VRFCoordinatorV2Mock', () => {
 
       let tx = await vrfCoordinatorV2Mock
         .connect(random)
-        .fulfillRandomWords(1, vrfConsumerV2.address)
+        ['fulfillRandomWords(uint256,address)'](1, vrfConsumerV2.address)
```
Comment (vreff) [inline contracts/test/v0.8/dev/VRFCoordinatorV2Mock.test.ts:257]:
```suggestion
        .fulfillRandomWords(1, vrfConsumerV2.address)
```

**Review by vreff — CHANGES_REQUESTED** [review]
Oops this is my bad! It looks like hardhat tests run into issues if you overload and call a function. So, to get around that without changing any tests we can just rename the 3 parameter function to `fulfillRandomWordsWithOverride`, so that we're no longer truly overloading and the tests are intact. You should be able to just commit the suggestions I've added, and I'll re-run the tests.

File: contracts/src/v0.8/mocks/VRFCoordinatorV2Mock.sol [hunk 5]
```diff
@@ -85,19 +85,41 @@ contract VRFCoordinatorV2Mock is VRFCoordinatorV2Interface {
...
+   * @notice consumer.
+   *
+   * @dev This mock uses a simplified formula for calculating payment amount and gas usage, and does
+   * @dev not account for all edge cases handled in the real VRF coordinator. When making requests
+   * @dev against the real coordinator a small amount of additional LINK is required.
+   *
+   * @param _requestId the request to fulfill
+   * @param _consumer the VRF randomness consumer to send the result to
+   * @param _words user provided random words
+   */
```
Comment (vreff) [inline contracts/src/v0.8/mocks/VRFCoordinatorV2Mock.sol:102]:
```suggestion
```

**Review by vreff — APPROVED** [review]

**Comment by cuckflong** [comment]
great thanks!

**Comment by vreff** [comment]
Moving to https://github.com/smartcontractkit/chainlink/pull/6697, so all checks will pass.

**Pull request closed without merging**